src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1124 @@
+    // described above, and whose following items depend on which constant.
+    // For the block "go to x: (50) y: ((x position) of (dog))", that mapping
+    // might look something like this:
+    //
+    //   inputs: {
+    //     X: [INPUT_SAME_BLOCK_SHADOW, [4, 50]],
+    //     Y: [INPUT_DIFF_BLOCK_SHADOW, "some block id", [4, -50]]
+    //   }
+    //
+    // The arrays [4, 50] and [4, -50] represent the two shadow blocks these
+    // inputs hold (the latter obscured by the "of" block). The value 4 is a
+    // constant referring to a "math_number" - essentially, it is the type of
+    // the shadow contained within that input.
+    //
+    // It's worth noting that some shadows are serialized as
+    // actual blocks on the target's "blocks" dictionary, and referred to by
+    // ID; the inputs in "switch to costume (item (random) of (costumes))"
+    // follow this structure:
+    //
+    //   inputs: {
+    //     COSTUME: [INPUT_DIFF_BLOCK_SHADOW, "id 1", "id 2"]
+    //   }
+    //
+    // ...where id 2 is the ID of the obscured shadow block. Specific details
+    // on how shadow blocks are serialized and whether they're stored as
+    // arrays or referenced by block ID is described in serializeInputShadow.
+    // As far as the input mapping is concerned, all that matters is that the
+    // two formats may be interchanged with one another.
+    //
+    // Also note that there are a couple blocks (specifically the variable and
+    // list-contents getters) which are serialized altogether in the compressed
+    // much the same as a compressed shadow, irregardless of the type of the
+    // input they've been placed inside. This is because they're so common in
+    // a project, and do not have any inputs of their own - only a field to
+    // identify which variable or list the block corresponds to. The input
+    // mapping for the block "set x to (spawn x)" would look something like
+    // this:
+    //
+    //   inputs: {
+    //     X: [INPUT_DIFF_BLOCK_SHADOW, [12, "spawn x", "someId"], [4, 0]]
+    //   }
+    //
+    // ...where someId is the ID of the variable, and [4, 0] is the obscured
+    // shadow block, as usual.
+
+    const { block, getBroadcastId, getCustomBlockData, initialValues, inputEntries, stage, target } = options;
+
+    const blockData = newBlockData();
+
+    const result: {
+      inputs: sb3.Block["inputs"];
+      blockData: BlockData;
+    } = {
+      inputs: {},
+      blockData
+    };
+
+    // Just for debugging!
+    const entryKeys = Object.keys(inputEntries);
+    const fieldEntryKeys = Object.keys(sb3.fieldTypeMap[block.opcode] || {});
+    const missingEntries = Object.keys(inputs).filter(
+      inp => !(entryKeys.includes(inp) || fieldEntryKeys.includes(inp))
+    );
+    for (const key of missingEntries) {
+      warn(`Missing entry for input ${key} on ${block.opcode} (${block.id} in ${target.name})`);
+    }
```

Comment:
How expensive is this? Would it be beneficial to only do all this work if `warn` is passed?

Reply:
Oh, good point! Definitely.

Buuuuuuuut this particular debugging code should probably be removed, because it's only necessary for working out new blocks - AFAICT I've tested all the blocks supported by sb-edit already, so it's not needed anymore.